### PR TITLE
Implement GPU frustum culling.

### PR DESCRIPTION
**Pull request by pcwalton**
This commit implements opt-in GPU frustum culling, built on top of the infrastructure in https://github.com/bevyengine/bevy/pull/12773. To enable it on a camera, add the `GpuCulling` component to it. To additionally disable CPU frustum culling, add the `NoCpuCulling` component. Note that adding `GpuCulling` without `NoCpuCulling` *currently* does nothing useful. The reason why `GpuCulling` doesn't automatically imply `NoCpuCulling` is that I intend to follow this patch up with GPU two-phase occlusion culling, and CPU frustum culling plus GPU occlusion culling seems like a very commonly-desired mode.

Adding the `GpuCulling` component to a view puts that view into *indirect mode*. This mode makes all drawcalls indirect, relying on the mesh preprocessing shader to allocate instances dynamically. In indirect mode, the `PreprocessWorkItem` `output_index` points not to a `MeshUniform` instance slot but instead to a set of `wgpu` `IndirectParameters`, from which it allocates an instance slot dynamically if frustum culling succeeds. Batch building has been updated to allocate and track indirect parameter slots, and the AABBs are now supplied to the GPU as `MeshCullingData`.

A small amount of code relating to the frustum culling has been borrowed from meshlets and moved into `maths.wgsl`. Note that standard Bevy frustum culling uses AABBs, while meshlets use bounding spheres; this means that not as much code can be shared as one might think.

This patch doesn't provide any way to perform GPU culling on shadow maps, to avoid making this patch bigger than it already is. That can be a followup.

### Added

* Frustum culling can now optionally be done on the GPU. To enable it, add the `GpuCulling` component to a camera.
* To disable CPU frustum culling, add `NoCpuCulling` to a camera. Note that `GpuCulling` doesn't automatically imply `NoCpuCulling`.

### CODE REVIEW

**Comment by BD103** [comment]
This seems worthy of the release notes. Thoughts?

**Comment by pcwalton** [comment]
This should be ready for review now. I've rebased it on top of #12773.

File: examples/stress_tests/many_cubes.rs
Comment (alice-i-cecile) [inline examples/stress_tests/many_cubes.rs:62]: Can we model these options as an enum with `Culling::Cpu/Gpu/None`?

Comment (pcwalton) [reply examples/stress_tests/many_cubes.rs:62]: Well, it'd have to be four options: no culling, CPU culling only, GPU culling only, and both CPU and GPU culling.(The last one will matter once the GPU can do occlusion culling and the CPU can't.) So that's why I made it two bools.

Comment (pcwalton) [reply examples/stress_tests/many_cubes.rs:62]: Admittedly this PR puts us in kind of a weird state, with a GPU culling feature that almost never helps, but that's because this PR is really a foundation for further improvements (that are themselves blocked on other PRs, and would be large).

File: crates/bevy_pbr/src/render/mesh.rs
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh.rs:981]: Should this be a more explicit message or is it just impossible to happen in this PR?

File: crates/bevy_pbr/src/render/mesh_preprocess.wgsl
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh_preprocess.wgsl:91]: This means a sphere is inside the frustum right? At first I interpreted "intersects" as if it's at an intersection between outside and inside the frustum so just making sure.

**Review by IceSentry — COMMENTED** [review]
This all looks broadly correct. I left a few comments but I'm not approving yet because I'm in the process of testing everything and making sure everything works.

File: crates/bevy_render/src/view/visibility/mod.rs
Comment (IceSentry) [inline crates/bevy_render/src/view/visibility/mod.rs:485]: Maybe it's just me but those double negative are a bit confusing. Would it make sense to use `!gpu_culling` instead? I guess there would still the the `!no_frustum_culling` so that might be even more confusing to use something different after.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (IceSentry) [reply crates/bevy_pbr/src/render/mesh.rs:981]: If I run many_cubes with no flag I always hit this path

File: crates/bevy_pbr/src/render/gpu_preprocess.rs
Comment (JMS55) [inline crates/bevy_pbr/src/render/gpu_preprocess.rs:69]: Not a new change, but why is Read<T> used here? What is it even?

**Comment by Elabajaba** [comment]
This is currently broken on dx12 for me. All the instances render with the same transform.

Probably just https://github.com/gfx-rs/wgpu/issues/4564

File: crates/bevy_pbr/src/render/mesh_preprocess.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_preprocess.wgsl:91]: It means that there's at least one point that's simultaneously inside the sphere and inside the frustum.

File: crates/bevy_render/src/view/visibility/mod.rs
Comment (pcwalton) [reply crates/bevy_render/src/view/visibility/mod.rs:485]: Well, "no CPU culling" doesn't imply "GPU culling". You can have GPU culling and CPU culling active at the same time. This isn't useful now, but once we have GPU occlusion culling, it can be useful. It might even be the default, depending on whether GPU occlusion culling or retained bins lands first. That's because CPU culling is very useful to reduce the amount of CPU load in `queue_material_meshes`, which will continue to be expensive at least until retained bins lands.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: That could happen if there are no meshes at all. Fixed.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: I merged it and `MeshCullingData`.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: Probably not. I reworded the comment.

File: crates/bevy_pbr/src/render/gpu_preprocess.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/gpu_preprocess.rs]: Yes, fixed.

File: crates/bevy_render/src/view/visibility/mod.rs
Comment (IceSentry) [reply crates/bevy_render/src/view/visibility/mod.rs:485]: > "no CPU culling" doesn't imply "GPU culling".

Right, yeah, I didn't consider that

**Comment by IceSentry** [comment]
> This is currently broken on dx12 for me

Since it's an opt-in feature anyway, I'm fine with merging it without DX12 support.

File: crates/bevy_pbr/src/render/mesh_preprocess.wgsl
Comment (IceSentry) [reply crates/bevy_pbr/src/render/mesh_preprocess.wgsl:91]: Yeah, makes sense. That's what I assumed, but wasn't completely sure. I originally read it as intersecting one of the frustum plane instead of being inside the volume of the frustum. Anyway, you can resolve this.

**Comment by Elabajaba** [comment]
> > This is currently broken on dx12 for me
> 
> Since it's an opt-in feature anyway, I'm fine with merging it without DX12 support.

It's an upstream wgpu bug (https://github.com/gfx-rs/wgpu/issues/2471), and it's now gated on the `VERTEX_AND_INSTANCE_INDEX_RESPECTS_RESPECTIVE_FIRST_VALUE_IN_INDIRECT_DRAW` downlevel flag, so once it's fixed upstream it should just start working in dx12.

File: crates/bevy_render/src/camera/camera.rs
Comment (IceSentry) [inline crates/bevy_render/src/camera/camera.rs:926]:
```suggestion
            if gpu_culling && *gpu_preprocessing_support == GpuPreprocessingSupport::Culling {
	            warn_once!("GpuCulling not supported on this platform");
```

Could be useful to have a warning if the feature is requested but not supported instead of silently ignoring it.

**Review by IceSentry — APPROVED** [review]
@mockersf would be nice to have a full rendering test run on this PR

Other than that LGTM. I did a few test and everything seemed to work as expected.

File: crates/bevy_render/src/camera/camera.rs
Comment (pcwalton) [reply crates/bevy_render/src/camera/camera.rs]: Added it.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl
Comment (superdump) [inline crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:71]: Was this name changed from sphere_intersects? If so, I'd prefer sphere_intersects for consistency with the main code. 'Sphere is inside' suggests the entire sphere is contained within, but that isn't what the frustum intersection means, right?

File: crates/bevy_pbr/src/render/gpu_preprocess.rs
Comment (superdump) [inline crates/bevy_pbr/src/render/gpu_preprocess.rs:83]: Is there a reason to not call this indirect? I'm guessing it will be because of the upcoming combinations you have mentioned elsewhere.

File: crates/bevy_pbr/src/render/gpu_preprocess.rs
Comment (superdump) [inline crates/bevy_pbr/src/render/gpu_preprocess.rs:102]: Is it worth making this is bitflags over a u8 so that if/when it grows, it grows using bitflags and not with additional members?

File: crates/bevy_pbr/src/render/mesh.rs
Comment (superdump) [inline crates/bevy_pbr/src/render/mesh.rs:434]: Why was inline removed?

File: crates/bevy_pbr/src/render/mesh.rs
Comment (superdump) [inline crates/bevy_pbr/src/render/mesh.rs:1263]: This sticks out and perhaps needs a comment - why?

File: crates/bevy_pbr/src/render/mesh_preprocess.wgsl
Comment (superdump) [inline crates/bevy_pbr/src/render/mesh_preprocess.wgsl:166]: For some reason I wonder if this is helpful on the GPU. It could be that all code has to be executed anyway, unless all waves go this way?

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (superdump) [inline crates/bevy_render/src/batching/gpu_preprocessing.rs:178]: Maybe we could call this:
```suggestion
    pub vertex_or_index_count: u32,
```

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (superdump) [inline crates/bevy_render/src/batching/gpu_preprocessing.rs:190]: And maybe we could call this:
```suggestion
    pub base_vertex_or_first_instance: u32,
```

I know it clashes with the `first_instance` just below, but it still feels clearer.

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (superdump) [inline crates/bevy_render/src/batching/gpu_preprocessing.rs:231]: Just noting that I was going to suggest adding a comment that this checks for a DX12 issue, but that feels like the kind of comment that becomes outdated. I hope one can look at the documentation of the downlevel flag and get more and up to date information about it.

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (superdump) [inline crates/bevy_render/src/batching/gpu_preprocessing.rs:419]: Should `Some((input_index, None))` end up here?

File: examples/3d/3d_shapes.rs
Comment (superdump) [inline examples/3d/3d_shapes.rs:95]: I'm not sure about these additions to the `3d_shapes`, `lighting`, and `deferred_rendering`. Maybe a new example is in order to test the culling combinations specifically?

File: crates/bevy_render/src/render_phase/mod.rs
Comment (superdump) [inline crates/bevy_render/src/render_phase/mod.rs:477]: Could also be on native with GLES or so I suppose? I'd say comment the feature that is present or not, suggesting WebGL2 as an example. So maybe 'we have compute shaders' rather than 'we're not on WebGL2'

**Review by superdump — COMMENTED** [review]
This basically looks good. I left a few comments. When the ones that sound blocking are addressed, I'll approve and merge. Thanks for the PR! <3

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (IceSentry) [reply crates/bevy_render/src/batching/gpu_preprocessing.rs:231]: This solution also means it will automatically work once wgpu has fixed the issue without us needing to change any code.

File: examples/3d/3d_shapes.rs
Comment (IceSentry) [reply examples/3d/3d_shapes.rs:95]: Oh, yeah, I missed that. I think it should just be removed.

File: crates/bevy_render/src/render_phase/mod.rs
Comment (IceSentry) [reply crates/bevy_render/src/render_phase/mod.rs:477]: That comment was there before. This PR just moved things around. I also got surprised by the mention of webgl2 but that's not this PRs fault. Although, still worth updating it I think.

File: examples/3d/3d_shapes.rs
Comment (pcwalton) [reply examples/3d/3d_shapes.rs:95]: Oh, oops, I meant to remove those but forgot :)

File: crates/bevy_pbr/src/render/mesh_preprocess.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_preprocess.wgsl:166]: Oh, interesting, you're right, I bet this can be skipped.

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (pcwalton) [reply crates/bevy_render/src/batching/gpu_preprocessing.rs:419]: Yes, because if `CompareData` is `None`, then the entity is unbatchable, and therefore we should ignore it after breaking the batch. I added a comment to that effect.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs:1263]: The giant comment in `IndirectParameters` explains why. I'll add a cross reference.

File: crates/bevy_pbr/src/render/gpu_preprocess.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/gpu_preprocess.rs:83]: There will probably be GPU occlusion culling too, which can be toggled on and off as it has a performance cost. So it's reasonable to support GPU frustum culling but not GPU occlusion culling.

File: crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/meshlet/cull_meshlets.wgsl:71]: I think I wrote "inside" because "intersects a plane" is wrong: it's not whether the sphere intersects the plane, but whether it's in the positive half-space defined by the plane. I'll make it `sphere_intersects_plane_half_space`.

File: crates/bevy_pbr/src/render/gpu_preprocess.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/gpu_preprocess.rs]: Done.

File: crates/bevy_pbr/src/render/mesh.rs
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh.rs]: I added it back.

File: crates/bevy_pbr/src/render/mesh_preprocess.wgsl
Comment (pcwalton) [reply crates/bevy_pbr/src/render/mesh_preprocess.wgsl]: I skip that check now.

File: crates/bevy_render/src/batching/gpu_preprocessing.rs
Comment (pcwalton) [reply crates/bevy_render/src/batching/gpu_preprocessing.rs]: Done.

File: examples/3d/3d_shapes.rs
Comment (pcwalton) [reply examples/3d/3d_shapes.rs]: Removed.

File: crates/bevy_render/src/render_phase/mod.rs
Comment (pcwalton) [reply crates/bevy_render/src/render_phase/mod.rs]: I reworded the comment.

**Comment by pcwalton** [comment]
Comments addressed.

**Review by superdump — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1307 if you'd like to help out.

**Pull request merged**
